rclrs/src/parameter/value.rs

```diff
@@ -0,0 +1,173 @@
+use crate::rcl_bindings::*;
+use std::ffi::CStr;
+
+/// A parameter value.
+///
+/// Such a value can be specified in YAML format on the command line, or in a parameter file.
+/// For instance `--param foo:='[1, 2, 3]'` specfies an `IntegerArray` value for the `foo` parameter.
+#[derive(Clone, Debug, PartialEq)]
+pub enum ParameterValue {
+    /// A boolean value.
+    ///
+    /// YAML examples: `true`, `True`, `yes`, `y`.
+    Bool(bool),
+    /// An i64 value.
+    ///
+    /// YAML examples: `1`, `-30`, `0x7C`.
+    Integer(i64),
+    /// An f64 value.
+    ///
+    /// YAML examples: `2.0`, `8e-3`.
+    Double(f64),
+    /// A string.
+    ///
+    /// YAML examples: `""`, `"björk"`, `"42"`.
+    ///
+    /// Unquoted strings are also possible, but not recommended,
+    /// because they may be interpreted as another data type.
+    String(String),
+    /// An array of u8.
+    ///
+    /// YAML example: Not possible to specify as YAML.
+    ByteArray(Vec<u8>),
+    /// An array of booleans.
+    ///
+    /// YAML example: `[true, false, false]`.
+    BoolArray(Vec<bool>),
+    /// An array of i64.
+    ///
+    /// YAML example: `[3, 4]`.
+    IntegerArray(Vec<i64>),
+    /// An array of f64.
+    ///
+    /// YAML example: `[5.0, 6e2]`.
+    DoubleArray(Vec<f64>),
+    /// An array of strings.
+    ///
+    /// YAML example: `["abc", ""]`.
+    StringArray(Vec<String>),
+}
+
+impl ParameterValue {
+    // Panics if the rcl_variant_t does not have exactly one field set.
+    //
+    // This function is unsafe because it is possible to pass in an rcl_variant_t
+    // containing dangling pointers, or incorrect array sizes.
+    pub(crate) unsafe fn from_rcl_variant(var: &rcl_variant_t) -> Self {
+        let num_active: u8 = [
+            !var.bool_value.is_null(),
+            !var.integer_value.is_null(),
+            !var.double_value.is_null(),
+            !var.string_value.is_null(),
+            !var.byte_array_value.is_null(),
+            !var.bool_array_value.is_null(),
+            !var.integer_array_value.is_null(),
+            !var.double_array_value.is_null(),
+            !var.string_array_value.is_null(),
+        ]
+        .into_iter()
+        .map(u8::from)
+        .sum();
+        assert_eq!(num_active, 1);
```

Comment:
This assertion will fail on empty list parameters, but neither of the big 2 client libraries can handle empty lists either.

Reply:
No problem with failing on empty lists. We should get clarification from `rclcpp` and `rclpy` what the expected behavior should be.

Reply:
Yeah, I will try to find out more about this edge case. What I know is that launch_ros forbids empty lists because they're not supported (yet), so I think this is expected and we're definitely fine for now.